doc/arch/adr-002-delete-unpublished-documents-from-index.md

```diff
@@ -0,0 +1,90 @@
+  more recent data in the index.
+- A document is unpublished and then immediately republished. The republishing
+  message is processed first. The later, stale unpublishing message should be
+  ignored and the published document should remain in the search index.
+- A document is published and then immediately unpublished. The unpublishing
+  event is processed first. The later, stale publishing message should not
+  re-add the document to the search index. It should remain unpublished
+
+## Decision
+
+### Use Elasticsearch's external versioning
+
+Elasticsearch supports [external versioning][es-versioning] for documents. This
+means that each document includes a version number from an external system. We
+will use the publishing event number from the publishing API.
+
+When the version number is set, Elasticsearch uses optimistic locking to ensure
+that a document is never updated or deleted if the new version number is less
+than or equal to the existing version number.
+
+This will take care of some of the edge cases:
+
+- If two publishing events arrive out of order, the second stale event will be
+  ignored because the version number is lower. Result: the document in the index
+  ends up in the expected state, which is the most recent version.
+- If a republishing message arrives before an earlier unpublishing message, the
+  unpublishing will be ignored because its version number is lower. Result: the
+  document remains in the index as expected.
+
+[es-versioning]: https://www.elastic.co/blog/elasticsearch-versioning-support
+
+### Delete unpublished documents
+
+The simplest way to handle an unpublishing event is to delete the document from
+the index. Unfortunately, this could lead to a race condition which will leave
+the index in a bad state:
+
+- If a document is published and then immediately unpublished by the unpublished
+  message arrives first, it will initially be deleted from the index. But when
+  the stale republishing message arrives, it will incorrectly recreate the
+  document. Result: the document remains in the search index despite being
+  unpublished.
+
+### Add tombstone records to index: postponed
+
+We considered handling this by storing unpublished documents as tombstone
+records. These would be temporary: they would be cleaned up by reindexing tasks,
+but would be present for long enough to prevent this race condition:
+
+- The unpublishing message arrives first and saves the document with an
+  'unpublished' flag. The stale republishing message then arrives, but the
+  message is ignored because its version is lower than the unpublishing message.
+  Result: the tombstone record remains in the search index as expected.
+
+We have decided to postpone the implementation of tombstone records until we
+have time to work on it. In the meantime, this race condition may occur, but the
+sequence of events leading to it are expected to be rare.
+
+### Rely on reindexing to fix inconsistencies
```

Comment:
this feels like a statement rather than a decision and should potentially be included in the delete option as a note on it.